### ISSUE TITLE
no function clause matching in Better.Plug.CORS.do_match/4

### ISSUE THREAD

**Issue by linjunpop** [whatyouhide/corsica#13]
My application's name is `Better`, I cannot find anywhere is calling the `do_match` function.

Console output:

```shell
[error] #PID<0.372.0> running Better.Endpoint terminated
Server: better.dev:80 (http)
Request: GET /
** (exit) an exception was raised:
    ** (FunctionClauseError) no function clause matching in Better.Plug.CORS.do_match/4
        (better) web/plugs/cors.ex:10: Better.Plug.CORS.do_match(%Plug.Conn{adapter: {Plug.Adapters.Cowboy.Conn, :...}, assigns: %{}, before_send: [#Function<1.25687313/1 in Plug.Logger.call/2>, #Function<0.52873082/1 in Phoenix.LiveReloader.before_send_inject_reloader/1>], body_params: %{}, cookies: %Plug.Conn.Unfetched{aspect: :cookies}, halted: false, host: "better.dev", method: "GET", owner: #PID<0.372.0>, params: %{}, path_info: [], peer: {{127, 0, 0, 1}, 60602}, port: 80, private: %{phoenix_endpoint: Better.Endpoint, plug_session_fetch: #Function<1.34159075/1 in Plug.Session.fetch_session/1>}, query_params: %{}, query_string: "", remote_ip: {127, 0, 0, 1}, req_cookies: %Plug.Conn.Unfetched{aspect: :cookies}, req_headers: [{"host", "better.dev"}, {"connection", "keep-alive"}, {"cache-control", "max-age=0"}, {"accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8"}, {"upgrade-insecure-requests", "1"}, {"user-agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/48.0.2564.116 Safari/537.36"}, {"accept-encoding", "gzip, deflate, sdch"}, {"accept-language", "en-US,en;q=0.8,zh-CN;q=0.6,zh;q=0.4"}, {"x-forwarded-for", "127.0.0.1"}, {"x-forwarded-host", "better.dev"}, {"x-forwarded-server", "better.dev"}], request_path: "/", resp_body: nil, resp_cookies: %{}, resp_headers: [{"cache-control", "max-age=0, private, must-revalidate"}, {"x-request-id", "30pjp4n2ken55fqov6v92h19qqdmtopq"}], scheme: :http, script_name: [], secret_key_base: "nYhQLU5Uvdc+9hXfG3aG+C0Z2kYF3/Ga0GzpN70U+i20a7UDO+Y9W45TaIFhHnpx", state: :unset, status: nil}, "GET", [], "better.dev")
        (better) web/plugs/cors.ex:1: Better.Plug.CORS.plug_builder_call/2
        (better) lib/better/endpoint.ex:1: Better.Endpoint.phoenix_pipeline/1
        (better) lib/plug/debugger.ex:93: Better.Endpoint."call (overridable 3)"/2
        (better) lib/phoenix/endpoint/render_errors.ex:34: Better.Endpoint.call/2
        (plug) lib/plug/adapters/cowboy/handler.ex:15: Plug.Adapters.Cowboy.Handler.upgrade/4
        (cowboy) src/cowboy_protocol.erl:442: :cowboy_protocol.execute/4
```

CORS module:

```elixir
# web/plugs/cors.ex

defmodule Better.Plug.CORS do
  use Corsica.Router,
    max_age: 600,
    allow_credentials: true,
    allow_headers: ~w(Authorization X-Device-ID X-Revision-ID Accept Content-Type User-Agent Origin Cache-Control Keep-Alive If-Modified-Since X-CSRF-Token),
    allow_methods: ~w(OPTIONS GET POST PUT DELETE),
    origins: Application.get_env(:better, :cors_origins),
    log: :debug

  resource "/api/auth_token"
end
```

Endpoint:

```elixir
defmodule Better.Endpoint do
  use Phoenix.Endpoint, otp_app: :better

  socket "/socket", Better.UserSocket

  # Serve at "/" the static files from "priv/static" directory.
  #
  # You should set gzip to true if you are running phoenix.digest
  # when deploying your static files in production.
  plug Plug.Static,
    at: "/", from: :better, gzip: false,
    only: ~w(css fonts images js favicon.ico robots.txt)

  # Code reloading can be explicitly enabled under the
  # :code_reloader configuration of your endpoint.
  if code_reloading? do
    socket "/phoenix/live_reload/socket", Phoenix.LiveReloader.Socket
    plug Phoenix.LiveReloader
    plug Phoenix.CodeReloader
  end

  plug Plug.RequestId
  plug Plug.Logger

  plug Plug.Parsers,
    parsers: [:urlencoded, :multipart, :json],
    pass: ["*/*"],
    json_decoder: Poison

  plug Plug.MethodOverride
  plug Plug.Head

  plug Plug.Session, #...

  plug Better.Plug.CORS

  plug Better.Router
end
```

**Comment by linjunpop** [comment]
`mix.lock`:

```elixir
%{"base64url": {:hex, :base64url, "0.0.1"},
  "bureaucrat": {:hex, :bureaucrat, "0.1.2"},
  "connection": {:hex, :connection, "1.0.2"},
  "corsica": {:hex, :corsica, "0.4.0"},
  "cowboy": {:hex, :cowboy, "1.0.4"},
  "cowlib": {:hex, :cowlib, "1.0.2"},
  "db_connection": {:hex, :db_connection, "0.2.4"},
  "decimal": {:hex, :decimal, "1.1.1"},
  "ecto": {:git, "https://github.com/elixir-lang/ecto.git", "9c3ee7d2be842eb50810dc3d7b40418a778e3949", []},
  "fs": {:hex, :fs, "0.9.2"},
  "gettext": {:hex, :gettext, "0.10.0"},
  "guardian": {:hex, :guardian, "0.10.1"},
  "inflex": {:hex, :inflex, "1.5.0"},
  "jose": {:hex, :jose, "1.7.0"},
  "phoenix": {:hex, :phoenix, "1.1.4"},
  "phoenix_ecto": {:hex, :phoenix_ecto, "3.0.0-beta.2"},
  "phoenix_html": {:hex, :phoenix_html, "2.5.0"},
  "phoenix_live_reload": {:hex, :phoenix_live_reload, "1.0.3"},
  "plug": {:hex, :plug, "1.1.2"},
  "poison": {:hex, :poison, "1.5.2"},
  "poolboy": {:hex, :poolboy, "1.5.1"},
  "postgrex": {:hex, :postgrex, "0.11.1"},
  "ranch": {:hex, :ranch, "1.2.1"},
  "uuid": {:hex, :uuid, "1.1.3"}}

```

Elixir:

```
Erlang/OTP 18 [erts-7.2.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.2.3
```

**Comment by whatyouhide** [comment]
Hey @linjunpop, sorry for the bug and for the long time it took me to fix it :disappointed_relieved: A fix is now in master (6da6f7dc1f22ce196317db16a181a4c768e98532), you can try it by modifying your `:corsica` dependency to `{:corsica, github: "whatyouhide/corsica"}`. Let me know if it works, I'll release a new version in the next few days (I need to take care of some other stuff first :smiley:).
